src/Compression/CompressionCodecQuantile.cpp

```diff
@@ -0,0 +1,218 @@
+            break;
+        default:
+            throw Exception(ErrorCodes::ILLEGAL_CODEC_PARAMETER, "Float size for Quantile codec can be 4 or 8, given {}", static_cast<UInt32>(float_width));
+    }
+
+    if (compress_error != PcoError::PcoSuccess)
+        throw Exception(ErrorCodes::CANNOT_COMPRESS, "Cannot compress with Quantile codec");
+
+    return HEADER_SIZE + compressed_data_written;
+}
+
+void CompressionCodecQuantile::doDecompressData(const char * source, UInt32 source_size, char * dest, UInt32 uncompressed_size) const
+{
+    if (source_size < HEADER_SIZE)
+        throw Exception(ErrorCodes::CANNOT_DECOMPRESS, "Cannot decompress Quantile-encoded data. File has wrong header {}", source_size);
+
+    UInt8 src_float_width = source[0], src_level = source[1];
+    PcoError decompress_error;
+
+    switch (src_float_width)
+    {
+        case sizeof(Float32):
+            decompress_error = pco_simple_decompress(PCO_TYPE_F32, reinterpret_cast<const uint8_t *>(source + HEADER_SIZE), source_size - HEADER_SIZE, reinterpret_cast<uint8_t *>(dest), uncompressed_size);
+            break;
+        case sizeof(Float64):
+            decompress_error = pco_simple_decompress(PCO_TYPE_F64, reinterpret_cast<const uint8_t *>(source + HEADER_SIZE), source_size - HEADER_SIZE, reinterpret_cast<uint8_t *>(dest), uncompressed_size);
+            break;
+        default:
+            throw Exception(ErrorCodes::ILLEGAL_CODEC_PARAMETER, "Float size for Quantile codec can be 4 or 8, given {}", static_cast<UInt32>(src_float_width));
+    }
+
+    if (decompress_error != PcoError::PcoSuccess)
+        throw Exception(ErrorCodes::CANNOT_DECOMPRESS, "Cannot decompress with Quantile codec {} {}", static_cast<UInt32>(src_float_width), static_cast<UInt32>(src_level));
+}
+
+UInt32 CompressionCodecQuantile::getMaxCompressedDataSize(UInt32 uncompressed_size) const
+{
+    UInt32 float_count = (uncompressed_size + float_width - 1) / float_width;
+    UInt32 file_size = 0;
+    PcoError err;
+
+    switch (float_width)
+    {
+        case sizeof(Float64):
+            err = pco_file_size(PCO_TYPE_F64, float_count, &file_size);
+            break;
+        case sizeof(Float32):
+            err = pco_file_size(PCO_TYPE_F32, float_count, &file_size);
+            break;
+        default:
+            throw Exception(ErrorCodes::ILLEGAL_CODEC_PARAMETER, "Float size for Quantile codec can be 4 or 8, given {}", static_cast<UInt32>(float_width));
+    }
+
+    if (err != PcoError::PcoSuccess)
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Cannot get max compressed data size");
+
+    return HEADER_SIZE + file_size;
+}
+
+namespace
+{
+
+UInt8 getFloatBytesSize(const IDataType & column_type)
+{
+    if (!WhichDataType(column_type).isFloat())
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Quantile codec is not applicable for {} because the data type is not float",
```

Comment:
Why limit to floats? Pcodec handles signed and unsigned ints too.

Reply:
Initially, the task was to add this codec for FP data types: [58394](https://github.com/ClickHouse/ClickHouse/issues/58394), but I also see no reason not to add handling for integers. I will add ints handling a little bit later.